src/sfizz/LFO.cpp

```diff
@@ -0,0 +1,295 @@
+    std::array<float, config::maxLFOSubs> subPhases_ {{}};
+    std::array<float, config::maxLFOSubs> sampleHoldMem_ {{}};
+};
+
+LFO::LFO()
+    : impl_(new Impl)
+{
+    impl_->sampleRate_ = config::defaultSampleRate;
+    impl_->desc_ = &LFODescription::getDefault();
+}
+
+LFO::~LFO()
+{
+}
+
+void LFO::setSampleRate(double sampleRate)
+{
+    impl_->sampleRate_ = sampleRate;
+}
+
+void LFO::configure(const LFODescription* desc)
+{
+    impl_->desc_ = desc ? desc : &LFODescription::getDefault();
+}
+
+void LFO::start()
+{
+    Impl& impl = *impl_;
+    const LFODescription& desc = *impl.desc_;
+    const float sampleRate = impl.sampleRate_;
+
+    impl.subPhases_.fill(desc.phase0);
+    impl.sampleHoldMem_.fill(0.0f);
+
+    const float delay = desc.delay;
+    impl.delayFramesLeft_ = (delay > 0) ? static_cast<size_t>(std::ceil(sampleRate * delay)) : 0u;
+
+    const float fade = desc.fade;
+    impl.fadeInPole_ = (fade > 0) ? std::exp(-1.0 / (fade * sampleRate)) : 0.0f;
+    impl.fadeInMemory_ = 0;
+}
+
+template <>
+inline float LFO::eval<LFOWave::Triangle>(float phase)
+{
+    float y = -4 * phase + 2;
+    y = (phase < 0.25f) ? (4 * phase) : y;
+    y = (phase > 0.75f) ? (4 * phase - 4) : y;
+    return y;
+}
+
+template <>
+inline float LFO::eval<LFOWave::Sine>(float phase)
+{
+    float x = phase + phase - 1;
+    return 4 * x * (1 - std::fabs(x));
+}
+
+template <>
+inline float LFO::eval<LFOWave::Pulse75>(float phase)
+{
+    return (phase < 0.75f) ? +1.0f : -1.0f;
+}
+
+template <>
+inline float LFO::eval<LFOWave::Square>(float phase)
+{
+    return (phase < 0.5f) ? +1.0f : -1.0f;
+}
+
+template <>
+inline float LFO::eval<LFOWave::Pulse25>(float phase)
+{
+    return (phase < 0.25f) ? +1.0f : -1.0f;
+}
+
+template <>
+inline float LFO::eval<LFOWave::Pulse12_5>(float phase)
+{
+    return (phase < 0.125f) ? +1.0f : -1.0f;
+}
+
+template <>
+inline float LFO::eval<LFOWave::Ramp>(float phase)
+{
+    return 2 * phase - 1;
+}
+
+template <>
+inline float LFO::eval<LFOWave::Saw>(float phase)
+{
+    return 1 - 2 * phase;
+}
+
+template <LFOWave W>
+void LFO::processWave(unsigned nth, absl::Span<float> out)
+{
+    Impl& impl = *impl_;
+    const LFODescription& desc = *impl.desc_;
+    const LFODescription::Sub& sub = desc.sub[nth];
+    const size_t numFrames = out.size();
+
+    float samplePeriod = 1.0f / impl.sampleRate_;
+    float phase = impl.subPhases_[nth];
+    float baseFreq = desc.freq;
+    float offset = sub.offset;
+    float ratio = sub.ratio;
+    float scale = sub.scale;
+
+    for (size_t i = 0; i < numFrames; ++i) {
+        out[i] += offset + scale * eval<W>(phase);
+
+        // TODO(jpc) lfoN_count: number of repetitions
+
+        float incrPhase = ratio * samplePeriod * baseFreq;
+        phase += incrPhase;
+        int numWraps = (int)phase;
+        phase -= numWraps;
+    }
+
+    impl.subPhases_[nth] = phase;
+}
+
+template <LFOWave W>
+void LFO::processSH(unsigned nth, absl::Span<float> out)
```

Comment:
These functions share alot of code, I wonder if we could avoid this repetition. I can try to work on this unless you have an idea.

Reply:
Repetitions, in particular `processWave`, it's not just for avoid the branch in the frame loop, also I wish for SSE impls of wave functions. All are vectorizable.